helix-core/src/test/java/org/apache/helix/controller/stages/TestRebalancePipeline.java

```diff
@@ -152,15 +152,15 @@ public void testDuplicateMsg() throws Exception {
     Assert.assertTrue(messages.get(0).getFromState().equalsIgnoreCase("SLAVE"));
     Assert.assertTrue(messages.get(0).getToState().equalsIgnoreCase("MASTER"));
 
-    runPipeline(event, dataRefresh);
-
+    Thread.sleep(MessageGenerationPhase.DEFAULT_OBSELETE_MSG_PURGE_DELAY);
     // Verify the stale message should be deleted
     Assert.assertTrue(TestHelper.verify(() -> {
+      runPipeline(event, dataRefresh);
```

Comment:
This changes the original test logic. We shall only need one pipeline run after the delay. And it shall guarantee the operation is done correctly.
If there is a concern about the delay time, we can wait for 2 * DEFAULT_OBSELETE_MSG_PURGE_DELAY.

Reply:
Ok. 

```
Thread.sleep(2 * MessageGenerationPhase.DEFAULT_OBSELETE_MSG_PURGE_DELAY);
runPipeline(event, dataRefresh);
Assert.assertTrue(TestHelper.verify(() -> {
if (dataCache.getStaleMessages().size() != 0) {
        return false;
      }
      return true;
    }, TestHelper.WAIT_DURATION));
```